hw/drivers/i2s/i2s_da1469x/src/i2s_da1469x.c

```diff
@@ -0,0 +1,433 @@
+{
+    int32_t denominator = 0;
+
+    uint32_t initial_deviation;
+    int32_t deviation;
+    int32_t minimum_deviation = INT32_MAX;
+    uint32_t bit_accumulator;
+    uint32_t frac_accumulator;
+    uint16_t fdiv = 1;
+
+    div->div = system_clock / bit_rate;
+    div->fdiv = 0;
+    bit_accumulator = bit_rate;
+
+    initial_deviation = system_clock - (bit_rate * div->div);
+    deviation = initial_deviation;
+    frac_accumulator = initial_deviation;
+
+    while (deviation != 0) {
+        if (deviation > 0) {
+            if (denominator == 16) {
+                break;
+            }
+            ++denominator;
+            fdiv <<= 1;
+            frac_accumulator += initial_deviation;
+        } else if (deviation < 0) {
+            fdiv |= 1;
+            bit_accumulator += bit_rate;
+        }
+        deviation = bit_accumulator - frac_accumulator;
+        if (abs(deviation) < minimum_deviation) {
+            minimum_deviation = abs(deviation);
+            div->fdiv = fdiv;
+        }
+    }
+}
+
+static void
+da1469x_i2s_dma_tx_isr(void *arg)
+{
+    struct i2s *i2s = da1469x_i2s.i2s;
+    struct i2s_sample_buffer *buffer;
+    (void)arg;
+
+    /* Change interrupt time to the end or half the circular buffer size, depending on which half is active. */
+    da1469x_i2s.dma_regs[1]->DMA_INT_REG = da1469x_i2s.dma_regs[1]->DMA_LEN_REG >> (da1469x_i2s.active_half);
+
+    /* DMA is already in other half, let active_half be consistent. */
+    da1469x_i2s.active_half ^= 1;
+
+    assert(da1469x_i2s.full_buffer_count > 0 && da1469x_i2s.full_buffer_count < 3);
+    da1469x_i2s.full_buffer_count--;
+    buffer = i2s_driver_buffer_get(i2s);
+    if (buffer) {
+        da1469x_i2s_fill_dma_buffer(buffer->sample_data, buffer->sample_count, i2s->sample_size_in_bytes);
+        i2s_driver_buffer_put(i2s, buffer);
+    } else {
+        da1469x_i2s.dma_regs[1]->DMA_CTRL_REG &= ~DMA_DMA0_CTRL_REG_DMA_ON_Msk;
+        if (da1469x_i2s.cfg.data_format != I2S_DATA_FRAME_16_16) {
+            da1469x_i2s.dma_regs[3]->DMA_CTRL_REG &= ~DMA_DMA0_CTRL_REG_DMA_ON_Msk;
+        }
+        i2s_driver_state_changed(i2s, I2S_STATE_OUT_OF_BUFFERS);
+    }
+}
+
+int
+i2s_driver_start(struct i2s *i2s)
+{
+    int rc = 0;
+    uint32_t system_clock;
+    uint32_t bit_rate;
+    uint32_t bits_per_frame;
+    struct i2s_sample_buffer *buffer;
+    struct da1469x_dma_buffer *dma_mem = da1469x_i2s.cfg.dma_memory;
+    struct pcm_div div;
+    const i2s_data_format_t data_format = da1469x_i2s.cfg.data_format;
+    struct da1469x_dma_config tx_cfg = {
+        .src_inc = true,
+        .dst_inc = false,
+        .priority = 0,
+        .burst_mode = 0,
+        .bus_width = data_format == I2S_DATA_FRAME_16_32 ? MCU_DMA_BUS_WIDTH_2B : MCU_DMA_BUS_WIDTH_4B,
+    };
+
+#if MYNEWT_VAL_CHOICE(MCU_SYSCLK_SOURCE, PLL96)
+    system_clock = 96000000;
+#else
+    system_clock = 32000000;
+#endif
```

Comment:
There's `SystemCoreClock` global defined by CMSIS which holds system core clock. In DA1469x MCU we do not have it defined properly, but perhaps it would be better to do so instead of hardcoding clocks here.